리액트 5주차 과제1

sinde530 opened the pull request:
강의 영상을 보면서 일단 작성 중입니다!

sinde530 committed:
App Test Code 작성

sinde530 committed:
App.test.code 음식 종류 테스트코드

sinde530 committed:
App.test.code 결과 리스트 반환하기

sinde530 committed:
App 컴포넌트 리스트 작성하기

sinde530 committed:
App.test.jsx refactoring

sinde530 committed:
redux settings

sinde530 committed:
action componenets create

sinde530 committed:
setCategories test code create

sinde530 committed:
categories test,component create

sinde530 committed:
카테고리 V표시 구현

sinde530 committed:
App test modify

sinde530 committed:
categories  test modify

sinde530 committed:
mock data create

sinde530 committed:
CategoriesContainer test create

sinde530 committed:
reducer test modify

sinde530 committed:
CategoriesContainer modify

sinde530 committed:
reducer test modify

sinde530 committed:
App Componenets modify

hannut91 commented on src/App.test.jsx:
```diff
@@ -0,0 +1,20 @@
+import { useSelector, useDispatch } from 'react-redux';
+
+import App from './App';
+import { CATEGORIES } from './fixture';
```
컴포넌트 임포트와 fixture임포트를 빈 줄로 구분해 주세요

sinde530 replied:
넵넵!

hannut91 commented on src/App.test.jsx:
```diff
@@ -0,0 +1,20 @@
+import { render } from '@testing-library/react';
+import { useSelector, useDispatch } from 'react-redux';
```
리액트 라이브러리 임포트와 테스팅 라이브러리 임포트를 빈 줄로 구분해 주세요

hannut91 commented on src/Categories.jsx:
```diff
@@ -0,0 +1,27 @@
+import { useState } from 'react';
+
+export default function Categories({ categories }) {
+  const [selected, setSelected] = useState();
+
+  const handleClickButton = ({ target: { name } }) => {
+    setSelected(name);
+  };
```
선택한 정보를 로컬 스테이트에 저장하고 있군요. 상태로 관리하지 않는 이유는 무엇인가요?

sinde530 replied:
차근차근 해보려고 합니다!

hannut91 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,18 @@
+    it('카테고리를 저장한다', () => {
+      const { categories } = reducer(previousState, setCategories(CATEGORIES));
+
+      expect(categories).toBe();
```
단정문을 완성해 주세요

sinde530 replied:
넵 수정하여 올렸습니다!

sinde530 committed:
categories api create

sinde530 committed:
App.test 코드리뷰 반영

sinde530 committed:
App Component api render

sinde530 committed:
App.test.jsx api add.

sinde530 committed:
api Regions add

sinde530 committed:
Categories에 카테고리가 존재하지 않다면 문구를 남긴다

hannut91 commented on src/App.jsx:
```diff
@@ -1,5 +1,26 @@
+  const loadCategories = async () => {
+    const response = await fetchCategories();
+    dispatch(setCategories(response));
+  };
```
컴포넌트에서 API호출 하는 일을 하고 있네요. 컴포넌트에서는 액션만 발행할 수 있습니다. 플럭스 아키텍처에 대해서 공부해 보세요

See also
  - https://facebook.github.io/flux/docs/in-depth-overview/

sinde530 replied:
넵 api폴더 따로 파서 넣어놨습니다!

hannut91 commented on src/Categories.jsx:
```diff
@@ -0,0 +1,29 @@
+    setSelected(name);
+  };
+
+  if (categories.length === 0) return <h2>카테고리가 존재하지 않습니다</h2>;
```
회사의 컨벤션에 따르는게 정답이지만, 보편적으로 한 줄이라도 중괄호를 쓰는게 좋을 것 같아요

sinde530 replied:
넵넵 감사합니다!

hannut91 commented on src/Categories.test.jsx:
```diff
@@ -0,0 +1,32 @@
+    <Categories categories={categories} />,
+  );
+
+  it('render', () => {
+    const { container } = renderComponent(CATEGORIES);
+
+    expect(container).toHaveTextContent('한식');
+  });
+
+  it('카테고리를 클릭하면 V표시가 된다', () => {
+    const { container, getByRole } = renderComponent();
+
+    const button = getByRole('button', { name: '한식' });
+    fireEvent.click(button);
+
+    expect(container).toHaveTextContent('한식(V)');
+  });
+
+  it('카테고리가 없으면 메세지가 출력된다', () => {
+    const { container } = renderComponent([]);
+
+    expect(container).toHaveTextContent('카테고리가 존재하지 않습니다');
+  });
```
`<Categories />`컴포넌트를 테스트를 먼저 작성하면, 카테고리 목록이 잘 보이는지, 카테고리 목록이 없는 경우에는 어떻게 되는지 테스트를 작성하실거에요.

그리고 나서 선택된 카테고리와 선택되지 않은 카테고리에 대해서 테스트를 작성하게 될탠데, 이 때 이 컴포넌트에 테스트가 점점 길어지고 테스트 대상이 많아서 복잡해진다고 느껴질 때가 있을거에요. 그 때가 컴포넌트를 분리해야겠다는 신호입니다.

sinde530 committed:
Categories 피드백 반영

`if (categories.length === 0) { return <h2>카테고리가 존재하지 않습니다</h2>; }`

sinde530 committed:
Categories.test.jsx Context Add.

sinde530 committed:
Library install

sinde530 committed:
api components Separate

sinde530 committed:
action.test add

sinde530 committed:
testcode All modify

sinde530 committed:
setRegions add.

sinde530 committed:
RegionsContainer create

sinde530 committed:
RegionsContainer.jsx create

sinde530 committed:
App.jsx modify

Uncaught Error: Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:

좋은 에러를 만났다.
- 함수 뒤에 괄호를 넣어주자

sinde530 committed:
Area Persenter Component Create

sinde530 committed:
카테고리,지역이 없을때 '존재하지 않다고'처리해준다

sinde530 committed:
선택된 카테고리 리덕스로 상태관리

hannut91 commented on src/Categories.jsx:
```diff
@@ -0,0 +1,25 @@
+    updateSelectedCategory(name);
+  };
+
+  if (!categories || categories.length === 0) { return <h2>카테고리가 존재하지 않습니다</h2>; }
```
```jsx
  const isEmpty = (arr) => (arr ?? []).length === 0;
  
  if (isEmpty(categories) { 
    return <h2>카테고리가 존재하지 않습니다</h2>; 
  }
```

if문은 위와 같이 내리는게 좋을 것 같습니다. 그리고 `isEmpty`같은 유틸함수를 사용하면 더 단순해지면서 의도도 명확하게 드러낼 수 있을 것 같아요. 혹은 lodash를 사용한다면 있는 것을 사용하셔도 됩니다.

아 그리고 categories가 없다면 `handleClickButton`은 사용되지 안잖아요. 그래서 핸들러 함수를 선언하기 전에 빠르게 종료하면 좋을 것 같네요

hannut91 commented on src/Categories.jsx:
```diff
@@ -0,0 +1,25 @@
+  if (!categories || categories.length === 0) { return <h2>카테고리가 존재하지 않습니다</h2>; }
+
+  return (
+    <>
```
여기 Fragment는 없어도 될 것 같아요

hannut91 commented on src/Categories.jsx:
```diff
@@ -0,0 +1,25 @@
+export default function Categories({ categories, selectedCategory, updateSelectedCategory }) {
```
지금은 선택됨을 `(V)`이걸로 보여주고 있잖아요. 그런데 우리가 이렇게 표시하지 않고 스타일로 표시하기로 요구사항이 변경되었다고 가정해 봅시다. 그러면 이 `<Categories />`컴포넌트를 수정해야해요. 이 컴포넌트는 카테고리 목록을 보여주고, 없을 경우 없다고 보여주는 역할을 하고 있는데, 선택됨을 표시하는 일도 하고 있죠. 따라서 선택됨을 표시해주는 역할을 하는 컴포넌트로 분리해 보시면 좋을 것 같아요

hannut91 commented on src/CategoriesContainer.jsx:
```diff
@@ -0,0 +1,15 @@
+  return (
+    <>
+      <Categories categories={categories} />
+    </>
```
여기도 Fragment는 없어도 될 것 같아요

sinde530 replied:
넵 강조되는부분들 열심히 기억해보겠습니다 ㅠㅠ

hannut91 commented on src/Regions.jsx:
```diff
@@ -0,0 +1,27 @@
+        <li key={id}>
+          <button type="button" name={name} onClick={handleClickButton}>
+            {
+              `${name}${selected === name ? '(V)' : ''}`
```
카테고리에서도 선택됨을 표시하는 일을 하고 있었는데 여기서도 하고 있군요. 지금은 만약 선택됨을 표시하는 방법이 바뀌면 둘 다 바꿔줘야 하겠군요

hannut91 commented on src/RegionsContainer.jsx:
```diff
@@ -0,0 +1,15 @@
+  }));
+
+  return (
+    <>
+      <Regions regions={regions} />
+    </>
```
여기도 Fragment는 없어도 될 것 같아요

sinde530 replied:
감사합니다!

hannut91 commented on src/Categories.test.jsx:
```diff
@@ -0,0 +1,47 @@
+
+describe('Categories', () => {
+  const renderComponent = (categories, selectedCategory) => render(
+    <Categories
+      categories={categories}
+      updateSelectedCategory={updateSelectedCategory}
+      selectedCategory={selectedCategory}
+    />,
+  );
+
+  it('render', () => {
+    const { container } = renderComponent(CATEGORIES, '');
+
+    expect(container).toHaveTextContent('한식');
+  });
+
+  context('카테고리가 있을 때', () => {
+    it('카테고리를 클릭하면 updateSelectedCategory가 호출된다', () => {
+      const { getByRole } = renderComponent(CATEGORIES, '');
+
+      const button = getByRole('button', { name: '한식' });
+      fireEvent.click(button);
+      expect(updateSelectedCategory).toBeCalled();
+    });
+
+    it('선택된 카테고리에는 V표시가 붙는다', () => {
+      const { container } = renderComponent(CATEGORIES, '한식');
+
+      expect(container).toHaveTextContent('한식(V)');
+    });
+  });
+
+  context('카테고리가 없을때', () => {
+    it('카테고리가 존재하지 않다는 문구가 출력된다', () => {
+      const { container } = renderComponent([]);
+
+      expect(container).toHaveTextContent('카테고리가 존재하지 않습니다');
+    });
+  });
+});
```
```js
describe('Categories', () => {
  given('categories', () => CATEGORIES);
  given('selectedCategory', () => undefined);

  const renderComponent = () => render(
    <Categories
      categories={given.categories}
      updateSelectedCategory={given.updateSelectedCategory}
      selectedCategory={selectedCategory}
    />,
  );

  context('선택된 카테고리가 있을 때', () => {
    given('selectedCategory', () => '한식');

    it('선택된 카테고리에는 V표시가 붙는다', () => {
      const { container } = renderComponent();

      expect(container).toHaveTextContent('한식(V)');
    });

    describe('카테고리를 클릭하면', () => {
      it('카테고리를 클릭하면 updateSelectedCategory가 호출된다', () => {
        const { getByRole } = renderComponent();

        const button = getByRole('button', { name: '한식' });
        fireEvent.click(button);

        expect(updateSelectedCategory).toBeCalled();
      });
    });
  });

  context('카테고리가 없을때', () => {
    given('categories', () => []);

    it('카테고리가 존재하지 않다는 문구가 출력된다', () => {
      const { container } = renderComponent();

      expect(container).toHaveTextContent('카테고리가 존재하지 않습니다');
    });
  });
});
```

given2 라이브러리를 사용하면 값을 lazy하게 할당할 수 있어서 무언가 변화하는 것을 명시할 때 의도를 드러내기가 쉽고 테스트도 단순하게 만들 수 있습니다. 한 번 사용해 보세요.

https://github.com/tatyshev/given2

sinde530 committed:
선택된 지역을 리덕스에서 관리하는 테스트 추가

sinde530 committed:
선택된 지역을 리덕스에서 관리

hannut91 commented:
과제 풀이 영상을 참고해서 과제를 완성해 주세요. 풀이 과정을 보면서 어떤 결정들을 하는지 아주 자세히 봐야 합니다.

sinde530 committed:
강의 따라가기

sinde530 committed:
loadInitiaData action에 만들어주기.

sinde530 committed:
reducer test 추가

sinde530 committed:
식당 목록 조회

sinde530 committed:
식당 조회시 2개를 받도록 테스트 구현

sinde530 committed:
식당 조회 컴포넌트 생성

sinde530 committed:
식당목록 컨테이너 컴포넌트 구현

sinde530 committed:
카테고리 이름 => id로 변경

sinde530 committed:
id,name으로 선택된 옵션을 저장하도록 테스트 수정
